x/oob/writer.go

```diff
@@ -0,0 +1,119 @@
+package oob
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+type oobWriter struct {
+	conn        *net.TCPConn
+	resetTTL    sync.Once
+	setTTL      sync.Once
+	oobPosition int64
+	sd          SocketDescriptor
+	oobByte     byte // Byte to send as OOB
+	disOOB      bool // Flag to enable disOOB mode
+	delay       time.Duration
+}
+
+var _ io.Writer = (*oobWriter)(nil)
+
+type oobWriterReaderFrom struct {
+	*oobWriter
+	rf io.ReaderFrom
+}
+
+var _ io.ReaderFrom = (*oobWriterReaderFrom)(nil)
+
+// NewWriter creates an [io.Writer] that sends an OOB byte at the specified "oobPosition".
+// If disOOB is enabled, it will apply the --disOOB strategy.
+// "oobByte" specifies the value of the byte to send out-of-band.
+func NewWriter(
+	conn *net.TCPConn,
+	sd SocketDescriptor,
+	oobPosition int64,
+	oobByte byte,
+	disOOB bool,
+	delay time.Duration,
+) io.Writer {
+	return &oobWriter{conn: conn, sd: sd, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB, delay: delay}
+}
+
+func (w *oobWriterReaderFrom) ReadFrom(source io.Reader) (int64, error) {
+	reader := io.MultiReader(io.LimitReader(source, w.oobPosition), source)
+	written, err := w.rf.ReadFrom(reader)
+	w.oobPosition -= written
+	return written, err
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if w.oobPosition > 0 && w.oobPosition < int64(len(data)) {
+		firstPart := data[:w.oobPosition+1]
```

Comment:
This will trigger index out of bounds if `len(data) == oobPosition+1`.

You may consider using https://linux.die.net/man/2/sendmsg, which supports the io vector, so you can pass two buffers for an atomic write.

Reply:
fixed